connman/src/service.c

```diff
@@ -510,10 +512,49 @@ static inline char *service_path(const char *ident)
 	return g_strconcat(CONNMAN_PATH, "/service/", ident, NULL);
 }
 
+
+static void count_available_service_type(struct connman_service *service,
+								bool add)
+{
+	GHashTable *table;
+	enum connman_service_type type;
+
+	if (!service)
+		return;
+
+	type = service->type;
+	table = service_type_hash[type];
+
+	DBG("%p/%s %s %c 1", service, service->name,
+					__connman_service_type2string(type),
+					add ? '+' : '-');
+
+	if (add) {
+		if (!is_available(service)) {
+			DBG("service %p/%s not available", service,
+							service->identifier);
+			return;
+		}
+
+		if (!table)
+			table = service_type_hash[type] = g_hash_table_new(
+						g_str_hash, g_str_equal);
+
+		g_hash_table_replace(table, service->identifier, service);
+	} else if (table) {
+		g_hash_table_remove(table, service->identifier);
+	}
+
+	DBG("%s service count: %d", __connman_service_type2string(type),
+					table ? g_hash_table_size(table) : 0);
+}
+
 static void service_remove(struct connman_service *service)
 {
 	service_list = g_list_remove(service_list, service);
 	g_hash_table_remove(service_hash, service->identifier);
+	g_hash_table_remove(service_type_hash[service->type],
+							service->identifier);
```

Comment:
Noting that this does not guard agains service_type_hash[service->type] == NULL - why not use count_available_service_type(service, false)?

Reply:
Yep, that is better and I moved it to be done as first here to be on the safe side with 97c3b643978731057463a21799e8bd362fb4a261